src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,406 @@
+     * @param pathBuilder          The path builder to use.
+     */
+    public ResourceCacheManager(final boolean supportsReloading, final ForgeConfigSpec.BooleanValue indexOffThreadConfig, final BiFunction<PackType, String, Path> pathBuilder)
+    {
+        this.supportsReloading = supportsReloading;
+        this.indexOffThreadConfigOption = indexOffThreadConfig;
+        this.pathBuilder = pathBuilder;
+    }
+
+    public boolean shouldIndexOffThread()
+    {
+        return !getConfigValue(this.indexOffThreadConfigOption);
+    }
+
+    public static boolean shouldUseCache()
+    {
+        return getConfigValue(ForgeConfig.COMMON.cachePackAccess);
+    }
+
+    private static boolean getConfigValue(Supplier<Boolean> configValue)
+    {
+        // Yes we catch the early loading error on purpose.
+        // Both the indexing off-thread and if caching is enabled are configurable, so we need to catch the error.
+        // So when they are loaded early we preserve the default behaviour which is to not cache or index on thread.
+        try
+        {
+            return configValue.get();
+        } catch (IllegalStateException e)
+        {
+            return false;
+        }
+    }
+
+    /**
+     * Invoked to trigger indexing of a given namespaces resources.
+     *
+     * @param namespace The namespace to index.
+     */
+    public void index(final String namespace)
+    {
+        for (final PackType packType : PackType.values())
+        {
+            final PackTypeAndNamespace key = new PackTypeAndNamespace(packType, namespace); // Make a key.
+            if (managersByNamespace.containsKey(key) && !supportsReloading)
+            { // If we do not support reloading we just skip this.
+                return;
+            }
+
+            // Create a new manager, overriding the previous one if it exists.
+            this.managersByNamespace.put(key, new NamespacedResourceCacheManager(packType, namespace, this.shouldIndexOffThread(), pathBuilder, this::createWalkingStream));
+
+            // Index the inner manager (which will for sure exist, and will know the pack type and namespace already)
+            this.managersByNamespace.get(key).index();
+        }
+    }
+
+    /**
+     * Handling method to create a new stream of paths in the target directory, which can then be walked.
+     * The caller needs to close the stream manually.
+     *
+     * @param path The path to walk.
+     * @return A stream of paths which reside in the given path.
+     * @throws IOException Thrown when the path does not exist or reading of path information failed.
+     */
+    @SuppressWarnings("resource")
+    // We close this automatically later on when the result is actually used. This is just a factory.
+    private Stream<Path> createWalkingStream(final Path path) throws IOException
+    {
+        return Files.walk(path)
+                .filter(resourcePath -> !resourcePath.toString().endsWith(".mcmeta"));
```

Comment:
This was default code which was already in place. It has existed since the beginning of the way forge looks up resources without a comment as to why. So no clue, as I have no idea why it is the way it is.